### ISSUE TITLE
Potential failure with two or more consecutive spaces in a quoted kernel parameter value

### ISSUE THREAD

**Issue by nl6720** [archlinux/mkinitcpio#52]
In https://github.com/dracutdevs/dracut/issues/720#issuecomment-791485382, @haraldh said:

> `mkinitcpio` might fail, if you have two consecutive spaces in the `PARTLABEL` or in any kernel parameter, looking at https://github.com/archlinux/mkinitcpio/blob/master/init_functions#L209

----

Personally I can't confirm this issue, as `root="PARTLABEL=Arch  Linux"` (two spaces between the words) seems to work for me. :shrug:

**Comment by Foxboron** [comment]
```diff
diff --git a/test/test_parse_cmdline b/test/test_parse_cmdline
index ff855e4..9605c22 100755
--- a/test/test_parse_cmdline
+++ b/test/test_parse_cmdline
@@ -68,6 +68,7 @@ test_parse() {
     result=0
     while [ "$#" -gt 0 ]; do
       key=$1 expected_value=$2
+      echo "key:" $key "expected:" $expected_value
       shift 2

       assert $expect_fail "$key" "$expected_value" || result=$e_assertion_failure
@@ -224,6 +225,11 @@ test_parse 'foo=bar bareword bar="ba az"' \
   'bareword' 'y' \
   'bar' 'ba az'

+test_parse 'root="PARTLABEL=Arch Linux"'\
+  'root' 'PARTLABEL=Arch Linux'
+test_parse 'root="PARTLABEL=Arch  Linux"'\
+  'root' 'PARTLABEL=Arch  Linux'
+
 if [ "$failed" -eq 0 ]; then
   echo "PASS: ${0##*/test_} ($tests tests)"
   exit 0
```

Gives us:

<pre>
key: root expected: PARTLABEL=Arch Linux
key: root expected: PARTLABEL=Arch Linux
FAIL: root: expected='PARTLABEL=Arch  Linux', got='PARTLABEL=Arch Linux'
FAIL: 1 tests failed
test/test_parse_cmdline fail
make: *** [Makefile:88: check] Error 1
</pre>

**Comment by Foxboron** [comment]
Yeah, the space is lost in the for loop regardless of the parsing we do to detecting quoted values later on. So this needs a complete rework or we do the C-code approach dracut does as well.

Or we just declare double spaces as unsupported.

**Comment by Foxboron** [comment]
@earnestly gave this an attempt and produced an AWK script that can parse a subset of cmdline, with proper quoting.

<pre>
#!/bin/awk -f
# An attempt to parse a reasonable subset of /proc/cmdline, produces each
# parameter on its own line while handling quotes, their removal and preserving
# those which are escaped.

BEGIN{
    # Append a space so the main loop can deal with everything instead of
    # having to work the final iteration afterwards.
    cmdline = ENVIRON["e"] " "
    len = length(cmdline)

    # The idea is to use a mask for storing the "logic" in an array so that
    # subsequent loop(s) can chop out the relevant regions and deal with them
    # as needed without having to deal with parsing.  E.g.
    #
    #       foo="bar \"baz\" qux"   b'a''re'word  last=value
    #       1111"111111111111111"0001'1''11'1111001111111111
    #
    # The quotes are preserved in the mask so that they can be easily removed
    # later.
    for(i=1; i<=len; ++i){
        ch = substr(cmdline, i, 1)

        if((ch == "\"" || ch == "'") && prev != "\\"){
            quote = ch
            quoted = !quoted
        }

        if(ch == " " && !quoted)
            mask[i] = 0
        else if(quote)
            mask[i] = quote
        else
            mask[i] = 1


        quote = ""
        prev = ch
    }

    len = length(mask)
    w = 1 # watermark

    for(i=1; i<=len; ++i){
        if(!mask[i]){
            region = substr(cmdline, w, i-w)

            # Rebuild the parameter without the quotes captured in the
            # mask.
            param = ""

            for(j=1; j<=length(region); ++j){
                ch = substr(region, j, 1)

                # Map the local region to the corresponding location in the
                # global mask.
                m = mask[j-1+w]

                # Quote removal by skipping them when constructing the new
                # parameter.  The check made on '1' is needed as it can appear
                # in both the parameter and the mask.
                # As 1 in the mask is a positive signal the '1' it represents
                # can always be included.
                if(m != ch || ch == 1)
                    param = param ch
            }

            print param
            w = i + 1
        }
    }
}
</pre>

**Comment by Earnestly** [comment]
The idea was to "re-encode" the kernel parameters as a sequence of `key=values` separated by a newline so that the shell's read loop can do the parameter expansion and additional logic as normal without having to worry about surrounding context.
